### PR TITLE
[Quantization] - Layerwise reloading of Attention/KV quantized models

### PR DESCRIPTION
**Pull request by Josephasafg**
## Purpose

- Implements layerwise reloading of attention scale weights (`k_scale`, `v_scale`, `q_scale`, `prob_scale`) for models with quantized KV caches
- Fixes scalar weight counting in default_weight_loader so CopyCounter tracks them correctly
## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/model_loader/reload/layerwise.py [hunk 1]
```diff
@@ -208,19 +205,15 @@ def finalize_layerwise_processing(model: torch.nn.Module, model_config: ModelCon
 
         # Attention/MLA layers are processed after all other layers
         if isinstance(layer, (Attention, MLAAttention)):
```
Comment (kylesayrs) [inline vllm/model_executor/model_loader/reload/layerwise.py:212]: This is somewhat out of scope for this PR, but note that attention must process after other layers have processed. This means that, for models which have delayed processing (due to imperfect loadwise loading or padding), we'll have to separate the attention layers into a separate for loop.

Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/layerwise.py:212]: We can do this in a separate PR, but I wanted to call it out just in case you run into any issues

File: vllm/model_executor/model_loader/reload/layerwise.py [hunk 2]
```diff
@@ -208,19 +205,15 @@ def finalize_layerwise_processing(model: torch.nn.Module, model_config: ModelCon
...
-                raise NotImplementedError(
-                    "Layerwise loading of Q/K/V scale weights is not implemented yet"
-                )
-
+            if info.load_numel > 0 and info.kernel_tensors is not None:
+                # Reload with new scale weights from checkpoint
+                _place_kernel_tensors(layer, info)
+                _reload_attention_scales(layer, info)
+            elif info.load_numel > 0 or info.kernel_tensors is None:
+                _layerwise_process(layer, info)
```
Comment (kylesayrs) [inline vllm/model_executor/model_loader/reload/layerwise.py]: I think we should continue to have the error for layerwise/online loading. This should technically be disallowed (ValueError), since attention must always process after linears

Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/layerwise.py]: I don't think that calling `_layerwise_process` on attention layers is valid, since they have different patterns to trigger processing.

File: vllm/model_executor/model_loader/reload/layerwise.py [hunk 3]
```diff
@@ -208,19 +205,15 @@ def finalize_layerwise_processing(model: torch.nn.Module, model_config: ModelCon
...
-                )
-
-            elif info.kernel_tensors is None:
-                raise NotImplementedError(
-                    "Layerwise loading of Q/K/V scale weights is not implemented yet"
-                )
-
+            if info.load_numel > 0 and info.kernel_tensors is not None:
+                # Reload with new scale weights from checkpoint
+                _place_kernel_tensors(layer, info)
```
Comment (kylesayrs) [inline vllm/model_executor/model_loader/reload/layerwise.py]: Why is this step necessary if we're going to load the weights later? Is it just to materialize the layer?

File: vllm/model_executor/model_loader/reload/layerwise.py [hunk 4]
```diff
@@ -249,6 +242,34 @@ def finalize_layerwise_reload(*args, **kwargs):
...
+
+    quant_method.process_weights_after_loading(layer)
+
+    # Some quant methods replace buffers via assignment, leaving them on CPU
+    target_device = info.restore_device
+    for name, tensor in get_layer_tensors(layer).items():
+        if tensor.device != target_device:
+            data = tensor.data.to(target_device)
+            if isinstance(tensor, torch.nn.Parameter):
+                tensor.data = data
```
Comment (kylesayrs) [inline vllm/model_executor/model_loader/reload/layerwise.py]: This assignment will break the original kernel tensor references, right?

Does the normal layerwise_process flow not work for attention? ie materialize, load, process, copy into kernel tensors?

File: vllm/model_executor/model_loader/reload/layerwise.py (hunk 2 above)
Comment (Josephasafg) [reply vllm/model_executor/model_loader/reload/layerwise.py]: I agree with this. I originally thought about adding a comment noting this case shouldn't happen, but raising an error is better. Changed to a `ValueError`

File: vllm/model_executor/model_loader/reload/layerwise.py (hunk 3 above)
Comment (Josephasafg) [reply vllm/model_executor/model_loader/reload/layerwise.py]: Yes.
The layer is on meta here so I needed to get it off meta before `.data.copy_()` at the end. Is there an alternative here maybe?

File: vllm/model_executor/model_loader/reload/layerwise.py (hunk 4 above)
Comment (Josephasafg) [reply vllm/model_executor/model_loader/reload/layerwise.py]: I Tried using just `layerwise_process`, it produced incorrect results. Attention scales need specific initial values to detect which scales were loaded from the checkpoint.
I did however switched to `.data.copy_()` to preserve kernel tensor references per your suggestion.

File: vllm/model_executor/model_loader/reload/layerwise.py (hunk 1 above)
Comment (Josephasafg) [reply vllm/model_executor/model_loader/reload/layerwise.py:212]: I can add something like this in this PR already, it might be a small change (basically collect all attn layers and process them after the "main" loop
```python
for layer in model.modules():
     ...
     if isinstance(layer, (Attention, MLAAttention)):
         deferred_attn.append((layer, info))
         continue
        # process non-attention layers

 for layer, info in deferred_attn:
     _finalize_attention_layer(layer, info, model_config)
```
Are there other ordering dependencies beyond MLA's that I should be aware of?

Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/layerwise.py:212]: You can see all of the ordering dependencies in the original [process_weights_after_loading](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/model_loader/utils.py#L95) implementation. As far as I'm aware of these, are the only ordering rules.

FYI because of this ordering constraint, we we likely never support online quant attention (not that we'd really need to).

File: vllm/model_executor/model_loader/reload/layerwise.py (hunk 4 above)
Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/layerwise.py]: I see, totally makes sense, attention is very unique in that regard. This flow comes with the assumption that dtype/shapes never change during attention processing, but I think this is a fine assumption to make for now. Consider calling it out in a comment.

I think that the 100% surefire best way to do this would be to capture values in addition to metadata for particular layers (attention), but this comes with tradeoffs that we don't want to deal with now.

**Review by kylesayrs — APPROVED** [review]

File: vllm/model_executor/model_loader/weight_utils.py [hunk 5]
```diff
@@ -1320,8 +1320,8 @@ def default_weight_loader(param: torch.Tensor, loaded_weight: torch.Tensor) -> N
         if param.numel() == 1 and loaded_weight.numel() == 1:
             # Sometimes scalar values aren't considered tensors with shapes
             # so if both param and loaded_weight are a scalar,
-            # "broadcast" instead of copy
-            param.data.fill_(loaded_weight.item())
```
Comment (kylesayrs) [inline vllm/model_executor/model_loader/weight_utils.py:1368]: I'm not sure if this might introduce a regression.

File: vllm/model_executor/model_loader/reload/layerwise.py (hunk 1 above)
Comment (Josephasafg) [reply vllm/model_executor/model_loader/reload/layerwise.py:212]: Thanks for the pointer! I added a "deferred" approach which simply loops on all of the attn layers after all of the non attn layers.
Good to know about the online quant constraint.

File: vllm/model_executor/model_loader/weight_utils.py (hunk 5 above)
Comment (Josephasafg) [reply vllm/model_executor/model_loader/weight_utils.py:1368]: `.fill_()` dispatches through `aten.fill_`, which `CopyCounter` doesn't intercept - so the counter stays at 0 and the attention reload path gets skipped. `.copy_()` fixes the counting. Open to other approaches if you see a better one! 

Will the CI be able to catch any regressions? I ran most of the tests in fp8 and test_reload and they seemed to pass

Comment (kylesayrs) [reply vllm/model_executor/model_loader/weight_utils.py:1368]: Makes sense, copying and filling into a value are equivalent if both tensors are scalars.

File: tests/model_executor/model_loader/test_reload.py [hunk 6]
```diff
@@ -164,6 +164,40 @@ def test_reload_weights(base_model, mul_model, add_model, tp_size, vllm_runner):
...
+        enable_prefix_caching=False,
+        max_model_len=16,
+        max_num_seqs=1,
+    ) as llm:
+        llm.collective_rpc(
+            "update_config",
+            kwargs={"overrides": {"load_config": {"load_format": "auto"}}},
+        )
+        llm.collective_rpc("reload_weights", kwargs={"weights_path": model})
+        reloaded_perp = llm.generate_prompt_perplexity(["3 4 = 7"], mask=["3 4 ="])[0]
```
Comment (kylesayrs) [inline tests/model_executor/model_loader/test_reload.py]: I think you can just do a basic perplexity test, rather doing a separate reference run. This model isn't pretrained to do multiplication anyways

**Review by kylesayrs — APPROVED** [review]

File: tests/model_executor/model_loader/test_reload.py (hunk 6 above)
Comment (Josephasafg) [reply tests/model_executor/model_loader/test_reload.py]: I simplified the test by making a single call to the model after reload and checking perplexity result

**Review by kylesayrs — APPROVED** [review]

**Review by mgoin — APPROVED** [review]
